### ISSUE TITLE
liblto_plugin.dll not built

### ISSUE THREAD

**Issue by Trass3r** [crosstool-ng/crosstool-ng#1122]
Tried building a Canadian arm-gcc 8.2 from WSL (Ubuntu 18.10) for Win64 targeting baremetal ARM with the crosstool-ng master.

It does build but no matter what config I try there is only a static lto_plugin.a in the libexec folder.

So in general which options are the crucial ones to set for target LTO to work? I'll attach the log later.

**Comment by stilor** [comment]
Assigning to post-next-release.

**Comment by Novakov** [comment]
I ran into this issue today and after playing with `.build` I managed build `liblto_plugin.dll` by running `make` manually with `/usr/lib/gcc/x86_64-w64-mingw32/6.3-win32/` added to PATH. I have no idea how to implement this in `crosstool-ng` (or maybe this approach is totally wrong) but maybe it will be useful for someone smarter than me :)

**Comment by IKukhta** [comment]
same issue with canadian build HOST-x86_64-w64-mingw32/i686-unknown-linux-gnu
mingw compiler was built previously by crosstool-ng

any news on this issue? is there any workarounds?

**Comment by IKukhta** [comment]
I have found the reason of an issue how to fix it. 

The reason why liblto_plugin.dll is not created is that gcc is looking for library depends during the build and even if --static-stdc++ is set for this plugin it explicitly looking for dynamic stdc++ library in host compiler toolchain. The same time if mingw is built using ct-ng it is configured to build only static stdc++ library.

To fix an issue you need to either use mingw built with both static and dynamic stdc++ (configure ... --enable-static --enable-dynamic) or you can fix your .config file and add **CT_SHARED_LIBS=y**.

You should note that flag is not handled by crosstool-ng properly and you cannot create defconfig file with this option. You should inject this option in your config file. I will create patch for this asap, also I would recommend to fix sample for mingw toolchain to produce both static and dynamic runtime.

**Comment by nolange** [comment]
Well, I ran into this weird issue aswell. But I am not using crosstool-ng for building the Windows toolchain (using my own small script).
What happens is that some mingw versions seem to enable the `-fstack-protector` feature and this needs libssp-0.dll. I am not sure where the lookup/check for dynamic plugins happens but it seems that this dll is not found.
I managed to build the toolchain including lto-plugin by adding `-fno-stack-protector`.

**Comment by 4ntoine** [comment]
Anything i can try except CT_SHARED_LIBS=y?  It did not help for the AVR - see the linked issue for AVR

**Comment by 4ntoine** [comment]
What about `CT_BINUTILS_PLUGINS`? if i enabled it (and static toolchain = n) than i do have liblto_shared.so. But i'd like to have it static in order not to depend on android libs and versions

**Comment by jhol** [comment]
`lto-plugin` is a sub-module of gcc. The source code and autoconf/automake files for it are contained int the gcc `lto-plugin/` subdirectory. On Linux hosts, the module is linked as `liblto_plugin.so`, and on Windows hosts the module should be linked as `liblto_plugin.dll` dynamic library, however because libtool linkage partially fails, the `liblto_plugin.a` static library is produced instead with the following error message:

```
/usr/bin/bash ./libtool --tag=CC --tag=disable-static  --mode=link x86_64-host_w64-mingw32-gcc -Wall  -O2 -g -pipe -I/home/jhol/workspace/ct-ng/_build/arm-none-eabi-gcc-11.2.0/build/HOST-x86_64-w64-mingw32/arm-none-eabi/buildtools/complibs-host/include -D__USE_MINGW_ACCESS -Wc,-static-libgcc -no-undefined -bindir "/home/jhol/workspace/ct-ng/_build/arm-none-eabi-gcc-11.2.0/toolchain/opt/jhol/arm-none-eabi-gcc-11.2.0/bin" -module -avoid-version -bindir /home/jhol/workspace/ct-ng/_build/arm-none-eabi-gcc-11.2.0/toolchain/opt/jhol/arm-none-eabi-gcc-11.2.0/libexec/gcc/arm-none-eabi/11.2.0  -Wc,../libiberty/libiberty.a  -Xcompiler '-static-libstdc++' -Xcompiler '-static-libgcc' '-L/home/jhol/workspace/ct-ng/_build/arm-none-eabi-gcc-11.2.0/build/HOST-x86_64-w64-mingw32/arm-none-eabi/buildtools/complibs-host/lib' '-lstdc++' '-lm' '-Wl,--stack,12582912' -o liblto_plugin.la -rpath /home/jhol/workspace/ct-ng/_build/arm-none-eabi-gcc-11.2.0/toolchain/opt/jhol/arm-none-eabi-gcc-11.2.0/libexec/gcc/arm-none-eabi/11.2.0 lto-plugin.lo
*** Warning: linker path does not have real file for library -lstdc++.
*** I have the capability to make that library automatically link in when
*** you link to this library.  But I can only do this if you have a
*** shared version of the library, which you do not appear to have
*** because I did check the linker path looking for a file starting
*** with libstdc++ but no candidates were found. (...for file magic test)

*** Warning: libtool could not satisfy all declared inter-library
*** dependencies of module liblto_plugin.  Therefore, libtool will create
*** a static module, that should work as long as the dlopening
*** application is linked with the -dlopen flag.
libtool: link: x86_64-host_w64-mingw32-ar rc .libs/liblto_plugin.a .libs/lto-plugin.o
libtool: link: x86_64-host_w64-mingw32-ranlib .libs/liblto_plugin.a
libtool: link: ( cd ".libs" && rm -f "liblto_plugin.la" && ln -s "../liblto_plugin.la" "liblto_plugin.la" )
WARNING: liblto_plugin.la is static, not copying to ../gcc/liblto_plugin.la
```

`lto_plugin` does not contain any C++ code, and has no need to reference `libstdc++`, however because of the gcc `LDFLAGS` set by crosstool-ng, `-lstdc++` is present in libtool linker command line.

Sometimes crosstool-ng builds versioms of gcc with the `--with-static-standard-libraries` option specified. This causes gcc to add `-static-libstdc++ -static-libgcc` to the LDFLAGS during `./configure`. However, libtool doesn't understand these options so the gcc build system translates them into `-Xcompiler -static-libstdc++ -Xcompiler -static-libgcc` in libtool invocations, which libtool then ignores.

Even with `CT_CC_GCC_STATIC_LIBSTDCXX` disabled, crosstool-ng never builds `libstdc++.dll`, only `libstdc++.a`. Therfore, in cases where there is no `libstdc++.dll` available on the build system e.g. canadian builds, there is never a real dynamic library available to meet the needs of libtool.

The reason `-lstdc++` is included in the gcc `LDFLAGS` is because of the following code fragment in `scripts/build/cc/gcc.sh`:

```
    if [ "${build_staticlinked}" = "yes" ]; then
        core_LDFLAGS+=("-static")
        if [ "${CT_GCC_older_than_6}" = "y" ]; then
            host_libstdcxx_flags+=("-static-libgcc")
            host_libstdcxx_flags+=("-Wl,-Bstatic,-lstdc++")
            host_libstdcxx_flags+=("-lm")
        fi
        # Companion libraries are build static (eg !shared), so
        # the libstdc++ is not pulled automatically, although it
        # is needed. Shoe-horn it in our LDFLAGS
        # Ditto libm on some Fedora boxen
        core_LDFLAGS+=("-lstdc++")
        core_LDFLAGS+=("-lm")
    else
        if [ "${CT_CC_GCC_STATIC_LIBSTDCXX}" = "y" -a "${CT_GCC_older_than_6}" = "y" ]; then
            # this is from CodeSourcery arm-2010q1-202-arm-none-linux-gnueabi.src.tar.bz2
            # build script
            # INFO: if the host gcc is gcc-4.5 then presumably we could use -static-libstdc++,
            #       see http://gcc.gnu.org/ml/gcc-patches/2009-06/msg01635.html
            host_libstdcxx_flags+=("-static-libgcc")
            host_libstdcxx_flags+=("-Wl,-Bstatic,-lstdc++,-Bdynamic")
            host_libstdcxx_flags+=("-lm")
        fi
        # When companion libraries are build static (eg !shared),
        # the libstdc++ is not pulled automatically, although it
        # is needed. Shoe-horn it in our LDFLAGS
        # Ditto libm on some Fedora boxen
        core_LDFLAGS+=("-lstdc++")
        core_LDFLAGS+=("-lm")
    fi
```

There are two fragments like this - one in `do_gcc_core_backend` and the other in `do_gcc_backend`

It appears therefore that this workaround is back-firing in our case. This code was written by Bryan Hundven in `1ad439907` in 2010. Is the problem that this patch was designed to solve still relevant?

Deleting these lines, as per this patch appears to fix the issue in my case:
[0001-WIP-Remove-lstdc-from-gcc-LDFLAGS.patch.txt](https://github.com/crosstool-ng/crosstool-ng/files/8133504/0001-WIP-Remove-lstdc-from-gcc-LDFLAGS.patch.txt)

With this patch, I don't know if the `LDFLAGS` value is set correctly in all cases. How can we verify that it causes no regressions?

**Comment by cpackham** [comment]
@jhol thanks for the well researched write-up. In terms of what to do - care to turn your WIP patch into a proper PR (use some of your write-up as the commit message). We probably can't merge it before the 1.25 release, but at least it'll get run through the CI builds. Post 1.25 we could merge the PR and see what shakes out.

**Comment by jhol** [comment]
@cpackham sorry for the delay in getting back to you - I had to jump through a few coorporate hoops.

Here is a PR to fix the issue: https://github.com/crosstool-ng/crosstool-ng/pull/1704 . It seems to me that the code changes are reasonable - but it's impossible for me to tell whether there any regressions anywhere. Hopefully the CI process will make it clear whether we are on the right track.

**Comment by jhol** [comment]
With https://github.com/crosstool-ng/crosstool-ng/pull/1704 merged, the issue should now be fixed

**Comment by cpackham** [comment]
Fixed via #1704
